warehouse/query-core/src/main/java/datawave/query/jexl/DatawavePartialInterpreter.java

```diff
@@ -0,0 +1,887 @@
+            }
+        }
+        
+        // may call getState() on left or just set the numerical result
+        leftState.numeric = left;
+        return leftState;
+    }
+    
+    public Object visit(ASTMethodNode node, Object data) {
+        if (data instanceof State) {
+            State state = (State) data;
+            if (state.getSet() != null) {
+                return super.visit(node, state.getSet());
+            }
+        } else if (data instanceof FunctionalSet) {
+            if (isFunctionalSetIncomplete(incompleteFields, (FunctionalSet) data)) {
+                return new State(true, true, (FunctionalSet) data);
+            }
+        }
+        
+        if (data == null) {
+            return super.visit(node, FunctionalSet.emptySet());
+        } else {
+            return super.visit(node, data);
+        }
+    }
+    
+    /**
+     * Helper method that determines if a functional set contains a ValueTuple for an incomplete field
+     * 
+     * @param fields
+     *            a set of incomplete fields
+     * @param set
+     *            a FunctionalSet
+     * @return true if the FunctionalSet contains a ValueTuple for an incomplete field
+     */
+    private boolean isFunctionalSetIncomplete(Set<String> fields, FunctionalSet set) {
+        ValueTuple vt;
+        for (Object o : set) {
+            if (o instanceof ValueTuple) {
+                vt = (ValueTuple) o;
+                if (fields.contains(vt.getFieldName())) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+    
+    /**
+     * Generate a {@link State} from a JexlNode, and its value from a super call to {@link DatawaveInterpreter}
+     *
+     * @param node
+     *            the node
+     * @param scriptExecuteResult
+     *            the result of script execution
+     * @return a State object
+     */
+    private State getState(JexlNode node, Object scriptExecuteResult) {
+        boolean fromUnknown = isIncomplete(node);
+        boolean matched = isMatched(scriptExecuteResult) || scriptExecuteResult instanceof ValueTuple || fromUnknown;
+        
+        if (scriptExecuteResult instanceof ValueTuple) {
+            return new State(true, fromUnknown, new FunctionalSet(Collections.singleton(scriptExecuteResult)));
+        } else if (scriptExecuteResult instanceof Integer || scriptExecuteResult instanceof Long) {
+            // if we got a numeric script result this could be the result of ASTMult or ASTDiv node. In this case we
+            // did match and must persist the result as a numeric
+            return new State(true, fromUnknown, new FunctionalSet(), scriptExecuteResult);
+        }
+        return new State(matched, fromUnknown);
+    }
+    
+    /**
+     * Used to track matched state on a per-node basis.
+     */
+    public class State {
+        boolean matched;
+        boolean fromUnknown;
+        FunctionalSet set;
+        Object numeric;
```

Comment:
could these be combined to just Object value? Then provide convenience methods to getValue() which attempt to cast to Numeric and FunctionalSet, String, and Boolean? Might make this cleaner in the rest of the impl. It can be a bit messy to follow all the logic.